### ISSUE TITLE
09-k-means-housing.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/betaniammc-machine-learning#6]
Hola @Betaniammc , por aquí las issues!

## LO QUE HAS HECHO BIEN

- **Eliminación de duplicados**: Correcto con `drop_duplicates()`
- **Selección de columnas**: Correctamente seleccionas solo Longitude, Latitude y MedInc
- **Escalado con MinMaxScaler**: Bien aplicado antes de K-Means
- **6 clusters**: Correcto según el ejercicio (aunque el codo sugiere 8, bien justificado)
- **Método del codo**: Bien implementado con KElbowVisualizer
- **Visualización geográfica**: Excelente uso de geopandas y contextily para mostrar los clusters sobre mapa real de California
- **Análisis detallado de clusters**: Muy buena interpretación de cada cluster con coordenadas e ingresos
- **Comparación de modelos**: Excelente comparar múltiples modelos (RandomForest, XGBoost, AdaBoost, GradientBoosting, LightGBM)
- **Conclusiones**: Buenas observaciones sobre el rendimiento de cada modelo

---

### 1. Según enunciado NO divides train/test ANTES del clustering

El ejercicio pide explícitamente:
> "Asegúrate de dividir convenientemente el conjunto de datos en train y test... podrás utilizarlos para entrenar el algoritmo no supervisado y luego para hacer predicciones sobre puntos nuevos"

Tú aplicas K-Means sobre **todo el dataset** y luego divides:

```python
# Tu código: K-Means sobre todo df_scaled
model.fit(df_scaled)
df["cluster"] = model.labels_

# Y DESPUÉS divides
X_train, X_test, y_train, y_test = train_test_split(X, y, ...)
```

**DEBERÍA SER:**
```python
# Primero dividir
X_train, X_test = train_test_split(df_scaled, test_size=0.2, random_state=18)

# K-Means solo con train
model = KMeans(n_clusters=6, random_state=42)
model.fit(X_train)

# Predecir en ambos
train_clusters = model.predict(X_train)
test_clusters = model.predict(X_test)
```

---

### 2. Mejora. Falta el Paso 3: Predicción con test y visualización comparativa

El ejercicio pide:
> "Ahora utiliza el modelo entrenado con el conjunto test y añade los puntos al gráfico anterior para confirmar que la predicción es satisfactoria"

No muestras una comparación visual de train vs test para validar que las predicciones son coherentes.

---

### 3. No guardas los modelos 

El ejercicio pide:
> "Almacena **ambos modelos** en la carpeta correspondiente"

Falta guardar K-Means y el modelo supervisado elegido:

```python
import pickle

# Guardar K-Means
with open('../models/kmeans_housing.pkl', 'wb') as f:
    pickle.dump(model, f)

# Guardar el mejor modelo supervisado (ej: XGBoost)
with open('../models/xgb_housing.pkl', 'wb') as f:
    pickle.dump(xgb, f)

# También el scaler
with open('../models/scaler_housing.pkl', 'wb') as f:
    pickle.dump(min_max_scaler, f)
```

---

### 4. Apunte de linea no necesaria en fit con fit_transform ya hecho

```python
min_max_features = min_max_scaler.fit_transform(df[num_variables]) # haces esto entonces...
min_max_features = min_max_scaler.transform(df[num_variables])  # esta línea sobra
```

El `fit_transform` ya hace fit + transform, la segunda línea no es necesaria :) .

---

### 5. Mejora, añadir métricas adicionales del modelo supervisado

Para clasificación multiclase, además de accuracy, sería bueno ver:
```python
from sklearn.metrics import classification_report, confusion_matrix

print(classification_report(y_test, xgb_y_pred_test))
print(confusion_matrix(y_test, xgb_y_pred_test))
```

### 6. Imports no utilizados

Tienes imports que no usas:
- `make_blobs`
- `silhouette_score`, `davies_bouldin_score`
- `GaussianNB`, `MultinomialNB`, `BernoulliNB`

No es grave, pero limpia el código para dejar las importanciones y tu trabajo menos pesado :) 

----

Excelente Betania <3  Te he puesto anotaciones sobre el enunciado: seguir el flujo del ejercicio (split → K-Means en train → predict en test → visualizar comparación) y guardar los modelos y algunos apuntes por si quieres mejorarlo. El análisis de clusters está muy bien documentado! ¡Buen trabajo!
